generators/chipyard/src/main/scala/IOBinders.scala

```diff
@@ -71,45 +74,60 @@ object GetSystemParameters {
   case IOBinders => up(IOBinders, site) + (tag.runtimeClass.toString ->
       ((t: Any) => {
         t match {
           case system: T => fn(system)
-          case _ => Nil
+          case _ => (Nil, Nil)
         }
       })
   )
 })
 
 // This macro composes with previous matches on some Top mixin. This is useful for
 // annotation-like binders, since those can typically be composed
-class ComposeIOBinder[T](fn: => (T) => Seq[IOBinderTuple])(implicit tag: ClassTag[T]) extends Config((site, here, up) => {
+class ComposeIOBinder[T, S <: Data](fn: => (T) => (Seq[S], Seq[IOCell]))(implicit tag: ClassTag[T]) extends Config((site, here, up) => {
   case IOBinders => up(IOBinders, site) + (tag.runtimeClass.toString ->
       ((t: Any) => {
         t match {
-          case system: T => (up(IOBinders, site)(tag.runtimeClass.toString)(system)
-            ++ fn(system))
-          case _ => Nil
+          case system: T =>
+            val r = up(IOBinders, site)(tag.runtimeClass.toString)(system)
+            val h = fn(system)
+            (r._1 ++ h._1, r._2 ++ h._2)
+          case _ => (Nil, Nil)
         }
       })
   )
 })
 
-// DOC include end: IOBinders
-
-object AddIOCells {
-  /**
-   * Add IO cells to a SiFive GPIO devices and name the IO ports.
-   * @param gpios A Seq of GPIO port bundles
-   * @param genFn A callable function to generate a DigitalGPIOCell module to use
-   * @return Returns a tuple of (a 2D Seq of Analog IOs corresponding to individual GPIO pins; a 2D Seq of IOCell module references)
-   */
-  def gpio(gpios: Seq[GPIOPortIO], genFn: () => DigitalGPIOCell = IOCell.genericGPIO): (Seq[Seq[Analog]], Seq[Seq[IOCell]]) = {
-    gpios.zipWithIndex.map({ case (gpio, i) =>
+object BoreHelper {
+  def apply(name: String, source: Clock): Clock = {
+    val clock_io = IO(Output(Clock())).suggestName(name)
+    val clock_wire = Wire(Clock()).suggestName(s"chiptop_${name}")
+    dontTouch(clock_wire)
+    clock_wire := false.B.asClock // necessary for BoringUtils to work properly
+    BoringUtils.bore(source, Seq(clock_wire))
+    clock_io := clock_wire
+    clock_io
+  }
+}
+
+
+case object IOCellKey extends Field[IOCellTypeParams](GenericIOCellParams())
+
+class ClockedIO[T <: Data](gen: T) extends Bundle {
+  val clock = Output(Clock())
+  val bits = gen
+  override def cloneType: this.type = (new ClockedIO(DataMirror.internal.chiselTypeClone[T](gen))).asInstanceOf[this.type]
+}
+
+class WithGPIOCells extends OverrideIOBinder({
+  (system: HasPeripheryGPIOModuleImp) => {
+    val (ports2d, cells2d) = system.gpio.zipWithIndex.map({ case (gpio, i) =>
       gpio.pins.zipWithIndex.map({ case (pin, j) =>
         val g = IO(Analog(1.W)).suggestName(s"gpio_${i}_${j}")
-        val iocell = genFn().suggestName(s"iocell_gpio_${i}_${j}")
+        val iocell = system.p(IOCellKey).gpio().suggestName(s"iocell_gpio_${i}_${j}")
         iocell.io.o := pin.o.oval
```

Comment:
We dropped some of the gpio pin signals here. Perhaps we should augment the genericIOCells with `pue` and `ds`?

Reply:
When we discussed this previously those were deemed "Extra" signals that we'd add along with other non-logical signals (because they don't affect the 1/0 logical state of the IO)